src/index_runner/es_indexers/indexer_utils.py

```diff
@@ -83,6 +83,36 @@ def _get_tags(ws_info):
         return []
 
 
+def merge_default_fields(indexer_ret, defaults):
+    """
+    merge defualt fields into existing default fields (if they exist)
+    """
+    if 'doc' not in indexer_ret:
+        raise ValueError(f"indexer return data should have 'doc' dictionary {indexer_ret}")
+    for field in defaults:
```

Comment:
There is a lot of nesting here that we could probably prevent. Here is another approach we could that only has a couple levels of nesting:


```py

def merge2(indexer_ret, defaults):
    """
    Merge default fields into existing default fields (if they exist).
    """
    if 'doc' not in indexer_ret:
        raise ValueError(f"indexer return data should have 'doc' dictionary {indexer_ret}")
    # Convert every field in each dict to sets
    indexer_ret['doc'] = vals_to_sets(indexer_ret['doc'])
    defaults = vals_to_sets(defaults)
    # Merge all the sets inside each dict
    for key, val in defaults.items():
        orig_val = indexer_ret['doc'].get(key, set())
        # Merge the default values with the original values
        indexer_ret['doc'][key] = list(val.union(orig_val))
    # Clean up any extra lists/sets wrapping scalar vals
    for key, val in indexer_ret['doc'].items():
        if isinstance(val, set):
            val = list(val)
            indexer_ret['doc'][key] = val
        if len(val) == 1:
            indexer_ret['doc'][key] = val[0]
    return indexer_ret


def vals_to_sets(dict_obj: dict) -> dict:
    """
    Convert all vals in a dict to sets of vals.
    Returns a new dict and does not mutate `dict_obj`.
    """
    ret = dict(dict_obj)  # clone
    for key, val in dict_obj.items():
        ret[key] = set(val if isinstance(val, list) else [val])
    return ret

```

Reply:
i like this solution much better